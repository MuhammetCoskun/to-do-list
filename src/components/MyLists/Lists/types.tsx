import { ToDoListItemType } from "../../ShowList/ToDoList/types";

export type ListType = {
  id: string;
  title: string;
  todos: ToDoListItemType[];
};

export type ListsProps = {
  lists: ListType[];
  onChooseList: (arg: string) => void;
};
