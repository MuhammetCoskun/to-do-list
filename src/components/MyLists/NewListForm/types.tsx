export type NewListFormProps = {
  onAddList: (listName: string) => void;
};
