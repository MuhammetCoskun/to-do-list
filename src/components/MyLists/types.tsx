import { ListType } from "./Lists/types";

export type MyListsProps = {
  onChooseList: (arg: string) => void;
  onAddList: (title: string) => void;
  lists: ListType[];
};
