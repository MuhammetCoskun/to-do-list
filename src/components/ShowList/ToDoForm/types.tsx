export interface ToDoFormProps {
  onAddTodo: (todo: string) => void;
  onRemoveTodo: () => void;
  onRemoveList: () => void;
}
