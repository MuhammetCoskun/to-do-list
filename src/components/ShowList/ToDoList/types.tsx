import { ListType } from "../../MyLists/Lists/types";

export type ToDoListItemType = {
  id: string;
  text: string;
  createdAt: string;
  isDone: boolean;
  doneAt?: string;
};

export type ToDoListItemProps = {
  id: string;
  key?: string;
  text: string;
  onRemove: (id: string) => void;
};

export type ToDoListProps = {
  list: ListType;
  onRemove: (id: string) => void;
};
