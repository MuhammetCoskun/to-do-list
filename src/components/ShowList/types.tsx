import { ListType } from "../MyLists/Lists/types";

export type ShowListProps = {
  listToShow: ListType;
  onAddTodo: (text: string) => void;
  onRemoveTodo: (id: string) => void;
};
