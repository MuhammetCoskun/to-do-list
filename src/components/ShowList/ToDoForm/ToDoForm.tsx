import React, { useRef, useState } from "react";
import { ToDoFormProps } from "./types";

const ToDoForm: React.FC<ToDoFormProps> = ({
  onAddTodo,
  onRemoveTodo,
  onRemoveList,
}) => {
  const todoRef = useRef<HTMLInputElement>(null);
  const [inputIsValid, setInputIsValid] = useState<boolean>(true);

  const submitHandler = (event: React.FormEvent) => {
    event.preventDefault();
    const enteredTodo = todoRef.current!.value;
    if (enteredTodo?.trim().length === 0) {
      setInputIsValid(false);
      return;
    }
    setInputIsValid(true);
    onAddTodo(enteredTodo);
    todoRef.current!.value = "";
  };

  return (
    <div className={"task-form"}>
      <form onSubmit={submitHandler}>
        <input
          className="new todo"
          type="text"
          placeholder="Enter a new to do"
          aria-label="new to do name"
          ref={todoRef}
          data-testid="input"
        />
        <button
          aria-label="create new to do"
          data-testid="btn-create"
          className="btn create"
        >
          +
        </button>
      </form>
      <div className="delete-item-container">
        <button
          className="btn delete"
          onClick={onRemoveTodo}
          data-testid="btn-delete-todo"
        >
          Clear completed tasks!
        </button>
        <button
          className="btn delete"
          onClick={onRemoveList}
          data-testid="btn-delete-list"
        >
          Delete list!
        </button>
      </div>
      {!inputIsValid && (
        <p data-testid="paragraph">Please enter a valid to do..</p>
      )}
    </div>
  );
};

export default ToDoForm;
