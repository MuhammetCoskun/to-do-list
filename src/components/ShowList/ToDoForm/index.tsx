import React, { useRef, useState } from "react";
import { ToDoFormProps } from "./types";

const ToDoForm: React.FC<ToDoFormProps> = ({ onAddTodo }) => {
  const todoRef = useRef<HTMLInputElement>(null);
  const [inputIsValid, setInputIsValid] = useState<boolean>(true);

  const submitHandler = (event: React.FormEvent) => {
    event.preventDefault();
    const enteredTodo = todoRef.current!.value;
    if (enteredTodo?.trim().length === 0) {
      setInputIsValid(false);
      return;
    }
    setInputIsValid(true);
    onAddTodo(enteredTodo);
    todoRef.current!.value = "";
  };

  return (
    <div className={"task-form"}>
      <form onSubmit={submitHandler}>
        <input
          className="new todo"
          type="text"
          placeholder="Enter a new to do"
          aria-label="new to do name"
          ref={todoRef}
          data-testid="input"
        />
        <button
          aria-label="create new to do"
          data-testid="button"
          className="btn create"
        >
          +
        </button>
      </form>
      <div className="delete-item-container">
        <button className="btn delete">Clear completed tasks!</button>
        <button className="btn delete">Delete list!</button>
      </div>
      {!inputIsValid && (
        <p data-testid="paragraph">Please enter a valid to do..</p>
      )}
    </div>
  );
};

export default ToDoForm;
//TODO add style
