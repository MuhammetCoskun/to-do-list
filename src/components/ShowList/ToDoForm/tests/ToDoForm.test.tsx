import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ToDoForm from "../ToDoForm";
import { ToDoFormProps } from "../types";

const mockOnAddTodo = jest.fn();
const mockOnRemoveTodo = jest.fn();
const mockOnRemoveList = jest.fn();

const properties = {
  onAddTodo: mockOnAddTodo,
  onRemoveTodo: mockOnRemoveTodo,
  onRemoveList: mockOnRemoveList,
};

const customRender = (props: ToDoFormProps) => {
  render(<ToDoForm {...props} />);
};

describe("ToDoForm component", () => {
  test("renders correctly", () => {
    const { asFragment } = render(
      <ToDoForm
        onAddTodo={mockOnAddTodo}
        onRemoveTodo={mockOnRemoveTodo}
        onRemoveList={mockOnRemoveList}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("renders error message on invalid input", () => {
    customRender({
      ...properties,
    });
    const buttonElement = screen.getByTestId("btn-create");
    userEvent.click(buttonElement);
    expect(
      screen.getByText("Please enter a valid to do..")
    ).toBeInTheDocument();
  });
  test("should call onAddTodo callback when create button clicked it", () => {
    customRender({
      ...properties,
    });
    const inputElement = screen.getByTestId("input");
    const buttonElement = screen.getByTestId("btn-create");
    userEvent.type(inputElement, "Write Tests!");
    userEvent.click(buttonElement);
    expect(mockOnAddTodo).toHaveBeenCalledWith("Write Tests!");
  });
  test("should call onRemoveTodo callback when clear completed task button clicked", () => {
    customRender({
      ...properties,
    });
    const clearToDoButton = screen.getByTestId("btn-delete-todo");
    userEvent.click(clearToDoButton);
    expect(mockOnRemoveTodo).toBeCalled();
  });
  test("should call onRemoveList callback when delete list button clicked", () => {
    customRender({
      ...properties,
    });
    const clearListButton = screen.getByTestId("btn-delete-list");
    userEvent.click(clearListButton);
    expect(mockOnRemoveList).toBeCalled();
  });
});
