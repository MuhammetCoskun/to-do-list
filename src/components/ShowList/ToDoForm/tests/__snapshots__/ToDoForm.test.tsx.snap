// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoForm component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="task-form"
  >
    <form>
      <input
        aria-label="new to do name"
        class="new todo"
        data-testid="input"
        placeholder="Enter a new to do"
        type="text"
      />
      <button
        aria-label="create new to do"
        class="btn create"
        data-testid="button"
      >
        +
      </button>
    </form>
    <div
      class="delete-item-container"
    >
      <button
        class="btn delete"
      >
        Clear completed tasks!
      </button>
      <button
        class="btn delete"
      >
        Delete list!
      </button>
    </div>
  </div>
</DocumentFragment>
`;
